services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/api_client/_base.py

```diff
@@ -0,0 +1,141 @@
+import asyncio
+import functools
+import inspect
+import logging
+from typing import Any, Awaitable, Callable, Optional
+
+from httpx import AsyncClient, ConnectError, Response, TransportError
+from httpx._types import TimeoutTypes, URLTypes
+from tenacity._asyncio import AsyncRetrying
+from tenacity.retry import retry_if_exception_type
+from tenacity.stop import stop_after_attempt
+from tenacity.wait import wait_exponential
+
+from ._errors import (
+    ClientTransportError,
+    UnexpectedStatusError,
+    WrongReturnType,
+    _RetryRequestError,
+)
+
+logger = logging.getLogger(__name__)
+
+
+def retry_on_errors(
+    request_func: Callable[..., Awaitable[Response]]
+) -> Callable[..., Awaitable[Response]]:
+    """
+    raises:
+    - `ClientTransportError`
+    - `httpx.HTTPError`
+    """
+    assert asyncio.iscoroutinefunction(request_func)
+
+    @functools.wraps(request_func)
+    async def request_wrapper(zelf: "BaseHThinClient", *args, **kwargs) -> Response:
+        # pylint:disable=protected-access
+        try:
+            async for attempt in AsyncRetrying(
+                # waits 1, 4, 8 seconds between retries and gives up
+                stop=stop_after_attempt(zelf._request_max_retries),
+                wait=wait_exponential(min=1),
+                retry=retry_if_exception_type(_RetryRequestError),
+                reraise=True,
+            ):
+                with attempt:
+                    try:
+                        logger.debug("Calling retry_on_errors")
+                        response: Response = await request_func(zelf, *args, **kwargs)
+                        return response
+                    except ConnectError as e:
+                        assert e._request
+                        logger.info(
+                            "[%s/%s]Retry. Unexpected ConnectError while requesting '%s %s': %s",
+                            attempt.retry_state.attempt_number,
+                            zelf._request_max_retries,
+                            e._request.method,
+                            e._request.url,
+                            f"{e=}",
+                        )
+                        raise _RetryRequestError(e) from e
```

Comment:
In order to filter which exceptions to retry, you should instead list them all in tenacities ``retry_if_exception_type(...)``. There is no need to create a ``_RetryRequestError`` exceptions group.

Moroever, there is no need to use with statement either

```python

@functools.wraps(request_func)
@retry(
        # waits 1, 4, 8 seconds between retries and gives up
        stop=stop_after_attempt(zelf._request_max_retries),
        wait=wait_exponential(min=1),
        retry=retry_if_exception_type(ConnectError,),
        reraise=True,
):
async def request_wrapper(zelf: "BaseHThinClient", *args, **kwargs) -> Response:
   r: Response = await request_func(zelf, *args, **kwargs)
   return r
```

Reply:
This will not work since you cannot access `zelf._request_max_retries`.

Reply:
Even sonarcloud detects that this code is [too complex](https://sonarcloud.io/project/issues?resolved=false&types=CODE_SMELL&pullRequest=3121&id=ITISFoundation_osparc-simcore&open=AYGLZKZw_kfjCAbIvxMW)

Reply:
Thanks for pointing it out. Think I've managed to simplify it quite a bit now.